### PR TITLE
Display dialog when opening `vscode://` urls

### LINKED ISSUE #147870: Github.dev -- Notebook Continue Working On... doesn't do anything.
This may be somehow machine specific as I'm not able to reproduce it on a second Windows machine.

This repo is based on a "start-from-nothing-on-Windows" exercise, so maybe that has something to do with the inconsistency of being able to repro. 

0) On a new Windows machine, install Git, Python (from python.org), and then navigate to https://github.com/greazer/covid-19-data.
1) Press . to open github.dev.
2) Open SampleNbsAndScripts/myanalysis_plotly.ipynb.
3) Run the first cell
4) Choose to Continue Working On... > Clone repository...

Nothing at all happens.

### PR DESCRIPTION
**Pull request by joyceerhl**
If no handler is installed, opening a protocol url will no-op, so we want to guide the user to download and install VS Code

This PR fixes https://github.com/microsoft/vscode/issues/147870

### CODE REVIEW

File: src/vs/workbench/browser/window.ts [hunk 1]
```diff
@@ -190,6 +190,30 @@ export class BrowserWindow extends Disposable {
...
+					// We cannot know whether the protocol handler succeeded.
+					// Display guidance in case it did not, e.g. VS Code is not installed.
+					if (matchesScheme(href, Schemas.vscode) || matchesScheme(href, 'vscode-insiders')) {
+						const showResult = await this.dialogService.show(
+							Severity.Info,
+							localize('you can close this tab', "All done. You can close this tab now."),
+							[
+								localize('continue here', "Continue here"),
+								localize('try again', "Try again"),
+								localize('install vs code', "Install VS Code")
```
Comment (bpasero) [inline src/vs/workbench/browser/window.ts:203]: We cannot hardcode things in OSS. Names have to come from product service, URLs should be fwlinks (we probably have one already for the download).

File: src/vs/workbench/browser/window.ts [hunk 2]
```diff
@@ -190,6 +190,30 @@ export class BrowserWindow extends Disposable {
 				// handling explicitly to prevent the workbench from going down.
 				else {
 					this.lifecycleService.withExpectedShutdown({ disableShutdownHandling: true }, () => window.location.href = href);
+
+					// We cannot know whether the protocol handler succeeded.
+					// Display guidance in case it did not, e.g. VS Code is not installed.
+					if (matchesScheme(href, Schemas.vscode) || matchesScheme(href, 'vscode-insiders')) {
```
Comment (bpasero) [inline src/vs/workbench/browser/window.ts:196]: there is a URL protocol in product service too iirc, cannot be hardcoded.

File: src/vs/workbench/browser/window.ts [hunk 3]
```diff
@@ -190,6 +190,30 @@ export class BrowserWindow extends Disposable {
...
+								localize('install vs code', "Install VS Code")
+							],
+							{
+								cancelId: 0,
+								detail: localize('opened vs code desktop detail', "We tried opening VS Code on your computer.")
+							},
+						);
+
+						if (showResult.choice === 1) {
+							this.lifecycleService.withExpectedShutdown({ disableShutdownHandling: true }, () => window.location.href = href);
```
Comment (bpasero) [inline src/vs/workbench/browser/window.ts:212]: Maybe this line could be extracted to a function since it is being duplicated now.

**Review by bpasero — CHANGES_REQUESTED** [review]

**Review by bpasero — APPROVED** [review]
You might want to tweak what button is the default button, currently it looks like this:

image

I think I would not expect a dialog to have it's default button in the middle.

**Comment by joyceerhl** [comment]
On Windows:
image

On Mac/Linux:
image

**Pull request merged**